### PR TITLE
Yukine's turn

### PR DESCRIPTION
**Pull request by AndrewJWin**
Haaaaave fun.

### CODE REVIEW

File: commands/general/help.js [hunk 1]
```diff
@@ -1,55 +1,98 @@
 const Command = require(`${process.cwd()}/base/Command.js`);
-
+const {MessageEmbed} = require("discord.js");
```
Comment (MrJacz) [inline commands/general/help.js]: make it `{ MessageEmbed }`

File: commands/general/help.js [hunk 2]
```diff
@@ -1,55 +1,98 @@
...
+    const embed = new MessageEmbed()
+      .setTimestamp()
+      .setColor(message.guild.me.roles.highest.color || 0x00AE86)
+      .setFooter(`Requested by ${message.author.tag}`, message.author.avatarURL()); 
+    // Preload MessageEmbed.
+  
+    // Here we sort out categories in case the user did not provide an argument.
+    
+    let currentCategory = "";
+    const sorted = this.client.commands.array().sort((p, c) => p.help.category > c.help.category ? 1 :  p.help.name > c.help.name && p.help.category === c.help.category ? 1 : -1 );
```
Comment (MrJacz) [inline commands/general/help.js]: There is no reason to make this an array tbh

File: commands/general/help.js [hunk 3]
```diff
@@ -1,55 +1,98 @@
...
+      });
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+
+      message.channel.send({embed});
+    } else {
+      let lol = 0;
+      sorted.forEach(c => {
+        if (c.help.category.toLowerCase() == args[0]) {
```
Comment (MrJacz) [inline commands/general/help.js]: this could be one line

File: commands/general/help.js [hunk 4]
```diff
@@ -1,55 +1,98 @@
...
-        if (error.message === "Cannot send messages to this user") {
-          return message.reply("I cannot send you the commands message, as it appears you have **Direct Messages's** disabled.");
-        } else {
-          throw error;
-        }       
+          num = num + 1;
+        }
+      });
+    
+      if (num != 0) {
```
Comment (MrJacz) [inline commands/general/help.js]: `if (num) {`
is easier

File: commands/general/help.js [hunk 5]
```diff
@@ -1,55 +1,98 @@
...
+        //message.channel.send(`${title}\n\n${description}\n${output}`, {code:"asciidoc"});
+        embed.setTitle("Command category help")
+          .setDescription(`A list of commands in the ${args[0]} category.  (Total of ${num} commands in this category)\n\nTo get help on a specific command do \`${message.settings.prefix}help <command>\`\n\n${num > 10 && page === 1 ? `To view more commands do\` ${message.settings.prefix}help <category> 2\`` : "" }`)
+          .addField("Commands", output);
+
+        message.channel.send({embed});
```
Comment (MrJacz) [inline commands/general/help.js]: return here

File: commands/general/help.js [hunk 6]
```diff
@@ -1,55 +1,98 @@
...
+    if (this.client.commands.has(command) || this.client.commands.forEach(command => {if (command.conf.aliases.includes(command)) return true;})) {
+      command = this.client.commands.get(command);
       if (level < this.client.levelCache[command.conf.permLevel]) return;
-      message.channel.send(`= ${command.help.name} = \n${command.help.description}\ncategory     :: ${command.help.category}\ncost         :: ₲${parseInt(command.help.cost)}\nusage        :: ${command.help.usage}\naliases      :: ${command.conf.aliases.join(", ")}\ndetails      :: ${command.help.extended}`, {code:"asciidoc"});    }
+      embed.setTitle(`${args[0]} help`)
+        .addField("Command description", command.help.description)
+        .addField("Command usage", `\`${command.help.usage}\``)
+        .addField("Command aliases", command.conf.aliases.length == 0 ? "None" : command.conf.aliases.join(", ") );
+    
+      message.channel.send({embed});
```
Comment (MrJacz) [inline commands/general/help.js]: return here as well

**Review by MrJacz — CHANGES_REQUESTED** [review]

File: commands/general/help.js (hunk 5 above)
Comment (AndrewJWin) [reply commands/general/help.js]: No need for return.

File: commands/general/help.js (hunk 6 above)
Comment (AndrewJWin) [reply commands/general/help.js]: No need for return also.

File: commands/general/help.js (hunk 5 above)
Comment (MrJacz) [reply commands/general/help.js]: Keep to a code style of returning

File: commands/general/help.js (hunk 6 above)
Comment (MrJacz) [reply commands/general/help.js]: Keep to a code style of returning also.

File: commands/general/help.js (hunk 1 above)
Comment (YorkAARGH) [reply commands/general/help.js]: Ditto

Comment (AndrewJWin) [reply commands/general/help.js]: Whoops, missed that one, that was a fix.

File: commands/general/help.js (hunk 2 above)
Comment (AndrewJWin) [reply commands/general/help.js]: There is, sorted gets run into a .forEach()

File: commands/general/help.js [hunk 7]
```diff
@@ -1,55 +1,98 @@
...
+*/
 class Help extends Command {
   constructor(client) {
     super(client, {
       name: "help",
-      description: "Displays the commands for your level.",
-      usage: "help [command]",
+      description: "Get help on a command, command category, or a setting",
       category: "System",
-      extended: "This command will display all available commands for your permission level, with the additonal option of getting per command information when you run 'help <command name>'.",
```
Comment (YorkAARGH) [inline commands/general/help.js]: Why did you remove extended info?

Comment (AndrewJWin) [reply commands/general/help.js]: Didn't realize I removed that?

File: commands/general/help.js [hunk 8]
```diff
@@ -1,55 +1,95 @@
...
+          return `\n\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+        }
+      }).join("");
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+
+      return message.channel.send({embed});
+    } else {
+      let lol = 0;
```
Comment (DevYukine) [inline commands/general/help.js]: dont name a variable lol ....

File: commands/general/help.js [hunk 9]
```diff
@@ -1,55 +1,95 @@
...
+      sorted.forEach(c => {
```
Comment (DevYukine) [inline commands/general/help.js]: forEach is asyncrous means that its not blocking and this code could cause a race condition

File: commands/general/help.js [hunk 10]
```diff
@@ -1,55 +1,95 @@
...
-      
-      if (this.client.commands.has(command)) command = this.client.commands.get(command);
-      else if (this.client.aliases.has(command)) command = this.client.commands.get(this.client.aliases.get(command));
-      else return;
-      
-      if (!message.guild && command.conf.guildOnly === true) return;
+    }
+    // Show individual command's help.
+    let command = type;
+    if (this.client.commands.has(command) || this.client.commands.forEach(command => {if (command.conf.aliases.includes(command)) return true;})) {
```
Comment (DevYukine) [inline commands/general/help.js]: forEach returns void, use `some` method instead

File: commands/general/help.js [hunk 11]
```diff
@@ -1,55 +1,95 @@
...
+        }
+      });
+    
+      if (num) {
+        //message.channel.send(`${title}\n\n${description}\n${output}`, {code:"asciidoc"});
+        embed.setTitle("Command category help")
+          .setDescription(`A list of commands in the ${type} category.  (Total of ${num} commands in this category)\n\nTo get help on a specific command do \`${message.settings.prefix}help <command>\`\n\n${num > 10 && pg === 1 ? `To view more commands do\` ${message.settings.prefix}help <category> 2\`` : "" }`)
+          .addField("Commands", output);
+
+        return message.channel.send({embed});
```
Comment (DevYukine) [inline commands/general/help.js]: same as in line 47

File: commands/general/help.js [hunk 12]
```diff
@@ -1,55 +1,95 @@
...
+    if (this.client.commands.has(command) || this.client.commands.forEach(command => {if (command.conf.aliases.includes(command)) return true;})) {
+      command = this.client.commands.get(command);
       if (level < this.client.levelCache[command.conf.permLevel]) return;
-      message.channel.send(`= ${command.help.name} = \n${command.help.description}\ncategory     :: ${command.help.category}\ncost         :: ₲${parseInt(command.help.cost)}\nusage        :: ${command.help.usage}\naliases      :: ${command.conf.aliases.join(", ")}\ndetails      :: ${command.help.extended}`, {code:"asciidoc"});    }
+      embed.setTitle(`${type} help`)
+        .addField("Command description", command.help.description)
+        .addField("Command usage", `\`${command.help.usage}\``)
+        .addField("Command aliases", command.conf.aliases.length == 0 ? "None" : command.conf.aliases.join(", ") );
+    
+      return message.channel.send({embed});
```
Comment (DevYukine) [inline commands/general/help.js]: same as in line 47

File: commands/general/help.js [hunk 13]
```diff
@@ -1,55 +1,95 @@
...
+        if (currentCategory !== cat && !type) {
+          currentCategory = cat;
+          return `\n\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+        }
+      }).join("");
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+
+      return message.channel.send({embed});
```
Comment (DevYukine) [inline commands/general/help.js]: only a preferation but since 11.3 you can do .send(embed)

**Review by DevYukine — CHANGES_REQUESTED** [review]

File: commands/general/help.js [hunk 14]
```diff
@@ -1,55 +1,91 @@
...
+          currentCategory = cat;
+          return `\n\`${message.settings.prefix}help ${cat.toLowerCase()}\` | Shows ${cat} commands`;
+        }
+      }).join("");
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+    } else {
+      let n = 0;
+      await sorted.forEach(c => {
```
Comment (DevYukine) [inline commands/general/help.js]: dont await there for no reason

File: commands/general/help.js [hunk 15]
```diff
@@ -1,55 +1,91 @@
...
-        });
-        await message.channel.send("Please check your inbox for a list of my commands.");
-        await message.author.send(output, {code:"asciidoc", split: { char: "\u200b" }});
-      } catch (error) {
-        if (error.message === "Cannot send messages to this user") {
-          return message.reply("I cannot send you the commands message, as it appears you have **Direct Messages's** disabled.");
-        } else {
-          throw error;
-        }       
+          num = num + 1;
```
Comment (DevYukine) [inline commands/general/help.js]: i would use the unary ++ operator here to increasy by 1

File: commands/general/help.js [hunk 16]
```diff
@@ -1,55 +1,91 @@
...
+        }
+      }).join("");
+
+      embed.setDescription(description)
+        .addField("Categories", output);
+    } else {
+      let n = 0;
+      await sorted.forEach(c => {
+        if (c.help.category.toLowerCase() === type.toLowerCase()) {
+          n = n + 1;
```
Comment (DevYukine) [inline commands/general/help.js]: i would use the unary ++ operator here to increasy by 1

**Review by DevYukine — CHANGES_REQUESTED** [review]

File: commands/general/help.js (hunk 14 above)
Comment (AndrewJWin) [reply commands/general/help.js]: Ach, left in. Doesn't exist on my master

**Review by DevYukine — APPROVED** [review]
lgtm now
